core/jobs/batch_jobs/opportunity_management_jobs.py

```diff
@@ -43,19 +45,172 @@
+        self,
+        question_skill_link_models: List[List[
+                question_models.QuestionSkillLinkModel
+            ]]
+        ) -> int:
+        """Counts the number of unique question ids.
+
+        Args:
+            question_skill_link_models: list[[list[QuestionSkillLinkModel]].
+                2D array of QuestionSkillLinkModels.
+
+        Returns:
+            int. The number of unique question ids.
+        """
+
+        question_ids = [link.question_id
+                        for link_list in question_skill_link_models
+                        for link in link_list]
+        return len(set(question_ids))
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        question_skill_link_models = (
+            self.pipeline
+            | 'Get all non-deleted QuestionSkillLinkModels' >> (
+                ndb_io.GetModels(
+                    question_models.QuestionSkillLinkModel
+                    .get_all(include_deleted=False))
+                )
+            | 'Group QuestionSkillLinkModels by skill id' >>
+                beam.GroupBy(lambda n: n.skill_id)
+        )
+
+        skills = (
+            self.pipeline
+            | 'Get all non-deleted SkillModels' >> (
+                ndb_io.GetModels(
+                    skill_models.SkillModel.get_all(include_deleted=False)))
+            | 'Get skill from model' >> beam.Map(
+                skill_fetchers.get_skill_from_model)
+            | 'Group Skills by skill id' >> beam.GroupBy(lambda m: m.id)
+        )
+
+        skills_with_question_counts = (
+            {
+                'skills': skills,
+                'question_skill_links': question_skill_link_models
+            }
+            | 'Merge by skill id' >> beam.CoGroupByKey()
+        )
+
+        opportunities_results = (
+            skills_with_question_counts
+            | beam.Map(
+                lambda n:
+                self._create_skill_opportunity_model(
+                    opportunity_domain.SkillOpportunity( # type: ignore[no-untyped-call]
+                        skill_id=n[1]['skills'][0][0].id,
+                        skill_description=n[1]['skills'][0][0].description,
+                        question_count=self.count_unique_question_ids(
+                            n[1]['question_skill_links']
+                        ))))
+        )
+
+        unused_put_result = (
+            opportunities_results
+            | 'Filter the results with SUCCESS status' >> beam.Filter(
+                lambda result: result.is_ok())
+            | 'Fetch the models to be put' >> beam.FlatMap(
+                lambda result: result.unwrap())
+            | 'Add ID as a key' >> beam.WithKeys(lambda model: model.id)  # pylint: disable=no-value-for-parameter
+            | 'Allow only one item per key' >> (
+                beam.combiners.Sample.FixedSizePerKey(1))
+            | 'Remove the IDs' >> beam.Values()  # pylint: disable=no-value-for-parameter
```

Comment:
Why is this needed?

Reply:
Whoops my bad, I missed this when copying over from the other job in the file. I don't think this is necessary. Thanks for pointing it out!